R/validate.R

```diff
@@ -18,16 +18,84 @@
 #' @keywords check-forecasts
 #' @examples
 #' as_forecast(example_binary)
-#' as_forecast(example_quantile)
-as_forecast <- function(data, ...) {
+#' as_forecast(
+#'   example_quantile,
+#'   forecast_unit = c("model", "target_type", "target_end_date",
+#'                     "horizon", "location")
+#' )
+as_forecast <- function(data,
+                        ...) {
   UseMethod("as_forecast")
 }
 
 #' @rdname as_forecast
+#' @param observed (optional) Name of the column in `data` that contains the
+#' observed values. This column will be renamed to "observed".
+#' @param predicted (optional) Name of the column in `data` that contains the
+#' predicted values. This column will be renamed to "predicted".
+#' @param model (optional) Name of the column in `data` that contains the names
+#' of the models/forecasters that generated the predicted values.
+#' This column will be renamed to "model".
+#' @param forecast_unit (optional) Name of the columns in `data` (after
+#' renaming) that denote the unit of a single forecast.
+#' See [get_forecast_unit()] for details.
+#' @param quantile_level (optional) Name of the column in `data` that contains
+#' the quantile level of the predicted values. This column will be renamed to
+#' "quantile_level". Only applicable to quantile-based forecasts.
+#' @param sample_id (optional) Name of the column in `data` that contains the
+#' sample id. This column will be renamed to "sample_id". Only applicable to
+#' sample-based forecasts.
 #' @export
-as_forecast.default <- function(data, ...) {
+as_forecast.default <- function(data,
+                                observed = NULL,
+                                predicted = NULL,
+                                model = NULL,
+                                forecast_unit = NULL,
+                                quantile_level = NULL,
+                                sample_id = NULL,
+                                ...) {
+  # check inputs
+  data <- ensure_data.table(data)
+  assert_character(observed, len = 1, null.ok = TRUE)
+  assert_subset(observed, names(data), empty.ok = TRUE)
+
+  assert_character(predicted, len = 1, null.ok = TRUE)
+  assert_subset(predicted, names(data), empty.ok = TRUE)
+
+  assert_character(model, len = 1, null.ok = TRUE)
+  assert_subset(model, names(data), empty.ok = TRUE)
+
+  assert_character(quantile_level, len = 1, null.ok = TRUE)
+  assert_subset(quantile_level, names(data), empty.ok = TRUE)
+
+  assert_character(sample_id, len = 1, null.ok = TRUE)
+  assert_subset(sample_id, names(data), empty.ok = TRUE)
+
+  # rename columns
+  if (!is.null(observed)) {
```

Comment:
rather than doing it this way I was sort of assuming you would have the default be `observed = "observed"` etc and then rename by selection? Is there a reason you went this way?

Reply:
Aah.. I just didn't think about it! 

Can do that. Only thing is that there is a bit of a problem with `quantile_level` and `sample_id` - you can't specify both. So either those have to be `NULL` or we'd need to check for the existence of the column.

Reply:
I like the mechanic - I'm not sure I like the kinds of error messages it produces. The following is the result if we do an input check such as `assert_subset(predicted, names(data))`, just like before:
<img width="1274" alt="image" src="https://github.com/epiforecasts/scoringutils/assets/37978797/2f436ddd-4eca-4d76-a439-da84f3d083b3">

We can put in some extra work to make the message a bit prettier, but it doesn't strike me as very elegant either. The fact that we're always trying to rename something, even if the user didn't consciously specify an argument makes it a bit hard to come up with a good error message if the default arguments don't match what the user has. 
<img width="558" alt="image" src="https://github.com/epiforecasts/scoringutils/assets/37978797/2dde6448-b730-4e02-aeb8-8468e33faa36">

In terms of the error messages the "only rename something if the user actively specifies it and then validate the end result" seems more elegant. The error messages you would have gotten earlier when attempting to run something like the following
<img width="597" alt="image" src="https://github.com/epiforecasts/scoringutils/assets/37978797/024630a4-cb08-4686-9741-196237704d6e">
would have been `Error: Assertion on 'data' failed: Both columns `observed` and predicted` are needed.` which seems clearer to me.